keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md

```diff
@@ -0,0 +1,606 @@
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+During `kubeadm init`, a ConfigMap is written under `kube-system` called `kubelet-config-x.y`
+where `x` is the Kubernetes MAJOR version and `y` the MINOR version. It includes a KubeletConfiguration
+object Kind. The same ConfigMap is read during `kubeadm join` and downloaded to a file so that the joining
+kubelet can use it via `--config`. The access to this ConfigMap for joining nodes, is made possible
+via RBAC rules created during `kubeadm init` using a similar versioned naming format - Role
+`kubeadm:kubelet-config-x.y` and RoleBinding `kubeadm:kubelet-config-x.y`, both under `kube-system`.
+
+Simplify the naming of the default kubelet configuration ConfigMap and related RBAC rules that kubeadm manages
+to the format `kubelet-config`, which does not include a version in the name. Given the stored configuration
+object is versioned (has GroupVersionKind), it would be up to the running kubeadm instance that GETs
+it to determine if this GroupVersionKind is supported.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently the overall design is faulty as `kubeadm init`, despite supporting a version skew with the kubelet
+of N-1, writes a single ConfigMap `kubelet-config-x.y` where `x.y` is the version of the control plane
+and not the version of the kubelet on the host where `kubeadm init` is called.
+
+Later, during `join` operations the version `x.y` is determined by looking up the control plane version
+of the cluster, via the `kubernetesVersion` field of the `ClusterConfiguration` object stored
+in the `kube-system/kubeadm-config` ConfigMap.
+
+During `upgrade` a new `kubelet-config-x.y+1` ConfigMap and RBAC rules are written where `y+1` is the target
+MINOR upgrade version. The old `kubelet-config-x.y` are not cleaned up.
+
+Encoding the version in the API object names is redundant and creates unnecessary complications for both
+kubeadm maintainers and users that wish to modify the object in the cluster.
```

Comment:
How would we deal with version skew between the kubelet-config contents and kubelet reading it?

Reply:
the current implementation of kubeadm does not deal with that. the stored KubeletConfiguration GVK is checked if it is known by a version of kubeadm and if it's unknown and error is reported. so the source of truth is the GVK inside the config and not the version in the name of the ConfigMap.

we expect an old GVK in the ConfigMap to still work with newer kubelets for a few releases, but otherwise the current state is that users have to manually upconvert when the time comes and until the kubelet has something like `kubelet --convert foo.yaml` in which case kubeadm could do it for them.

Reply:
added some notes about that under "risk 3".